### PR TITLE
MGMT-8263: skipperize SRO repo

### PR DESCRIPTION
**Pull request by yevgeny-shnaidman**
This PR adds skipper configuration to the repo, without changing
the previous functionality. It will enable to run Makefile command
via skipper, which will eluminate need for any on-host installation of the packages

### CODE REVIEW

**Comment by ybettan** [comment]
> which will eluminate need for any on-host installation of the packages

nit:
eliminate

File: Dockerfile.special-resource-operator-build [hunk 1]
```diff
@@ -0,0 +1,17 @@
+FROM registry.ci.openshift.org/openshift/release:golang-1.16
+
+ENV GO111MODULE=on
+ENV GOFLAGS=""
+
+RUN curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3
+RUN chmod 700 get_helm.sh
+RUN ./get_helm.sh
+RUN yum install -y which
+RUN yum install -y podman docker
```
Comment (ybettan) [inline Dockerfile.special-resource-operator-build:10]: Is it working correctly? I know that in RHEL those 2 aren't working well together :) 
Why not `podman` and `podan-docker` instead?

File: Makefile.helper.mk [hunk 2]
```diff
@@ -4,11 +4,12 @@
 SPECIALRESOURCE  ?= driver-container-base
 NAMESPACE        ?= openshift-special-resource-operator
 PULLPOLICY       ?= IfNotPresent
-TAG              ?= $(shell git branch --show-current)
+TAG              ?= $(shell git rev-parse --abbrev-ref HEAD)
```
Comment (ybettan) [inline Makefile.helper.mk:7]: We should probably update the documentation as well in that case:
https://github.com/openshift-psap/special-resource-operator#from-the-cli

TL;DR

> Note: The image TAG will default to the name of the current git branch, but can be overriden by setting the TAG variable. See the Makefile for more details

File: skipper.env [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+DISABLED_HOST_VALIDATIONS
+CONTAINER_COMMAND
+DISABLED_STEPS
+MINIKUBE_HOME
+KUBECONFIG
+IMAGE
+VERSION
+REPO
+SPECIALRESOURCE
+IMG
```
Comment (ybettan) [inline skipper.env:42]: Do we have a way to get all the host's env variables mounted by default?

The way it is implemented now, for each new env variable to SRO we will need to update it here as well.

File: skipper.yaml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+registry: quay.io
+build-container-image: special-resource-operator-build
+
+containers:
+  special-resource-operator-build: Dockerfile.special-resource-operator-build
+volumes:
+    - $HOME/.cache/go-build:$HOME/.cache/go-build
+    - $HOME/go/pkg:/go/pkg
+    - $HOME/.cache/golangci-lint:$HOME/.cache/golangci-lint
+    - $HOME/.docker:$HOME/.docker
```
Comment (ybettan) [inline skipper.yaml:10]: I believe this is related to the Docker daemon socket? what happens if the machine doesn't contain `docker` but `podman` ?

File: skipper.env (hunk 3 above)
Comment (yevgeny-shnaidman) [reply skipper.env:42]: currently no, but we don't expect to have many new env variables. In assisted-service, since everybody used skipper, so every added ENV variable had to be added to skipper.env. in SRO, we will have to add it manually

File: skipper.yaml (hunk 4 above)
Comment (yevgeny-shnaidman) [reply skipper.yaml:10]: it should work as is. Docker needs socket since there is a docker daemon running. With podman this mapping will be unused

File: Makefile.helper.mk (hunk 2 above)
Comment (yevgeny-shnaidman) [reply Makefile.helper.mk:7]: you mean README?

**Comment by yevgeny-shnaidman** [comment]
> > which will eluminate need for any on-host installation of the packages
> 
> nit: eliminate

fixed

File: Dockerfile.special-resource-operator-build (hunk 1 above)
Comment (yevgeny-shnaidman) [reply Dockerfile.special-resource-operator-build:10]: i don't see any difference in those.This is the same setup we use in AI, and it works there

Comment (ybettan) [reply Dockerfile.special-resource-operator-build:10]: Probably there is no package conflicts in the base image.
I am OK with that anyway.

File: Makefile.helper.mk (hunk 2 above)
Comment (ybettan) [reply Makefile.helper.mk:7]: Yes, specifically the part I have link (inside README)

File: skipper.env (hunk 3 above)
Comment (ybettan) [reply skipper.env:42]: So there will be no difference between skipper in assisted and skipper in SRO :( 
Anyway, I don't have any better solution, just was wondering if it is possible.

File: skipper.env [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+ENABLE_KUBE_API
+STORAGE
+CRD_OPTIONS
+TWINE_USERNAME
+TWINE_PASSWORD
+TWINE_REPOSITORY
+GENERATE_CRD
+BUNDLE_IMAGE
+INDEX_IMAGE
+DEBUG_SERVICE
```
Comment (qbarrand) [inline skipper.env:30]: I think there are many variables that we are not using - `TWINE_*`, `DEBUG_*`, `LOCAL_ASSISTED_ORG`...

File: Dockerfile.special-resource-operator-build (hunk 1 above)
Comment (qbarrand) [inline Dockerfile.special-resource-operator-build:10]:
```suggestion
RUN yum install -y docker podman which
```

File: Dockerfile.special-resource-operator-build [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+RUN curl -L "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl" -o /tmp/kubectl
+RUN install -o root -g root -m 0755 /tmp/kubectl /usr/local/bin/kubectl
+RUN curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(go env GOPATH)/bin v1.42.1
+RUN GO111MODULE=on go install golang.stackrox.io/kube-linter/cmd/kube-linter@latest
```
Comment (qbarrand) [inline Dockerfile.special-resource-operator-build:14]:
```suggestion
RUN go install golang.stackrox.io/kube-linter/cmd/kube-linter@latest
```
`GO111MODULE` is already set line 3.

Comment (qbarrand) [reply Dockerfile.special-resource-operator-build:14]: Skipper would then run all `make` commands in the container, right? I am not sure then if it's worth installing those dependencies manually; the Makefile has [targets](https://github.com/openshift-psap/special-resource-operator/blob/master/Makefile.helper.mk#L45) that do it.

File: Makefile.helper.mk (hunk 2 above)
Comment (qbarrand) [reply Makefile.helper.mk:7]: What does this change bring @yevgeny-shnaidman? Does it handle tags better?

File: Dockerfile.special-resource-operator-build (hunk 6 above)
Comment (yevgeny-shnaidman) [reply Dockerfile.special-resource-operator-build:14]: skipper is just a wrap for any command that user would like to run on Makefile. So, if previously user would run:
make local-image-build
now he will run:
skipper make local-image-build
Regarding installing dependencies:  since skipper is actually a running image, then installing dependencies via Makefile means that they will need to be installed on every skipper run. Installing them via Dockerfile means that they are present in the image all the tiime

**Comment by yevgeny-shnaidman** [comment]
@qbarrand , since skipper image is based on golang.1.16 image, the git version there does not support --show-current option. The new TAG definition is exact equivalent of the previous definition

File: Dockerfile.special-resource-operator-build (hunk 1 above)
Comment (yevgeny-shnaidman) [reply Dockerfile.special-resource-operator-build:10]: fixed

File: skipper.env (hunk 5 above)
Comment (yevgeny-shnaidman) [reply skipper.env:30]: fixed. Left some variable that are currently unused, but we might need them later

File: Dockerfile.special-resource-operator-build (hunk 1 above)
Comment (eranco74) [reply Dockerfile.special-resource-operator-build:10]: Don't forget to add ` && yum -y clean all  && rm -rf /var/cache` while you are at it

File: skipper.yaml (hunk 4 above)
Comment (eranco74) [reply skipper.yaml:10]: The be accurate this isn't related to the docker daemon socket.
This volume is for mounting the auth file with container registries credentials (usually kept under `$HOME/.docker/config.json`).
It is useful and works for podman as well (though in podman the default path for the auth file is `$XDG_RUNTIME_DIR/containers/auth.json.`)
